### PR TITLE
Show silenced errors in separate tab

### PR DESCRIPTION
**Pull request by peterrehm**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #16740 |
| License | MIT |
| Doc PR | - |

Show the silenced errors in a separate tab other than the debug information to reflect also the counts of the toolbar. Also related to @javiereguiluz's comment here https://github.com/symfony/symfony/pull/10466#issuecomment-37808597.

bildschirmfoto 2015-11-30 um 16 46 09

bildschirmfoto 2015-11-30 um 16 47 05

### CODE REVIEW

**Comment by stof** [comment]
this will have to wait for 3.1. It is too late to incorporate such change in 2.8/3.0 IMO, given that the release is today

**Comment by peterrehm** [comment]
Would be helpful in case of silenced errors as I experienced them and it is not really a feature, just makes the errors more visible.

**Comment by jakzal** [comment]
:+1: nice touch

status: reviewed

**Comment by sstok** [comment]
:+1:

**Comment by peterrehm** [comment]
The question is just the target branch. As it is for me more like a bugfix it would be great to get this still into 2.8 to avoid confusion.

**Comment by Tobion** [comment]
I agree with this PR. But I'd like to see two changes:

- I would suggest to rename `Silenced Errors` with `Silenced Reports` (both in the toolbar and log panel).
The thing is that "errors" means it is a failure per se. But the silenced reports can also be warnings, info, notices, e.g. `@trigger_error('NOTICE', E_USER_NOTICE);` which is obviously not an error according according to the log levels.
- The actual level of the silenced report is not easily seeable in the silenced tab. Thus I would suggest to add it above the time in the time column. That basically just means converting `context.type` to the `E_...` constants. So where the `Info & Errors` tab shows the Log Level, we show the `E_...` constant names for silenced reports.

If we do above changes, I'd also vote :+1: to merge it in 2.8 as it will prevent confusion due to these changes later.

**Comment by peterrehm** [comment]
@Tobion How would you convert the type back to the constant names? Any idea about that?

**Comment by Tobion** [comment]
Similar to https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Debug/ErrorHandler.php#L49-L64 but with the constant names instead. Maybe the named constant should actually be added in the `context` directory. Then you do not need to do this in the view.

**Comment by peterrehm** [comment]
I have updated it to include an error_level in the context. This way we can keep the same renderer and I think this provides enough information. @Tobion what do you think?

File: src/Symfony/Component/HttpKernel/DataCollector/LoggerDataCollector.php [hunk 1]
```diff
@@ -22,6 +22,24 @@
  */
 class LoggerDataCollector extends DataCollector implements LateDataCollectorInterface
 {
+    private $errorLevel = array(
```
Comment (Tobion) [inline src/Symfony/Component/HttpKernel/DataCollector/LoggerDataCollector.php]: please rename to `$errorNames`

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/logger.html.twig [hunk 2]
```diff
@@ -53,10 +53,12 @@
         </div>
     {% else %}
         {# sort collected logs in groups #}
-        {% set deprecation_logs, debug_logs, info_and_error_logs = [], [], [] %}
+        {% set deprecation_logs, debug_logs, info_and_error_logs, silenced_logs = [], [], [], [] %}
         {% for log in collector.logs %}
             {% if log.context.level is defined and log.context.type is defined and log.context.type in [constant('E_DEPRECATED'), constant('E_USER_DEPRECATED')] %}
                 {% set deprecation_logs = deprecation_logs|merge([log]) %}
+            {% elseif log.priorityName == 'DEBUG' and log.context.scream is defined and log.context.scream == true  %}
```
Comment (Tobion) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/logger.html.twig]: the condition `log.priorityName == 'DEBUG'` can be removed as the log priority doen't change the fact that its a silenced error

File: src/Symfony/Component/HttpKernel/DataCollector/LoggerDataCollector.php [hunk 3]
```diff
@@ -106,6 +124,7 @@ private function sanitizeLogs($logs)
             if (isset($context['type'], $context['file'], $context['line'], $context['level'])) {
                 $errorId = md5("{$context['type']}/{$context['line']}/{$context['file']}\x00{$log['message']}", true);
                 $silenced = !($context['type'] & $context['level']);
+                $context['error_level'] = isset($this->errorLevel[$context['type']]) ? $this->errorLevel[$context['type']] : '';
```
Comment (Tobion) [inline src/Symfony/Component/HttpKernel/DataCollector/LoggerDataCollector.php]: I'd write this as follows to just name it `name` and have it as first key as it belong to the type:

```
if (isset($this->errorNames[$context['type']])) {
                    $context = array_merge(array('name' => $this->errorNames[$context['type']]), $context);
                }
```

**Comment by peterrehm** [comment]
@Tobion Updated accordingly

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/logger.html.twig [hunk 4]
```diff
@@ -53,10 +53,12 @@
         </div>
     {% else %}
         {# sort collected logs in groups #}
-        {% set deprecation_logs, debug_logs, info_and_error_logs = [], [], [] %}
+        {% set deprecation_logs, debug_logs, info_and_error_logs, silenced_logs = [], [], [], [] %}
         {% for log in collector.logs %}
             {% if log.context.level is defined and log.context.type is defined and log.context.type in [constant('E_DEPRECATED'), constant('E_USER_DEPRECATED')] %}
                 {% set deprecation_logs = deprecation_logs|merge([log]) %}
+            {% elseif and log.context.scream is defined and log.context.scream == true  %}
```
Comment (Tobion) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/logger.html.twig]: syntax error

**Comment by peterrehm** [comment]
@Tobion Updated.

**Comment by Tobion** [comment]
:+1:

**Comment by Tobion** [comment]
@javiereguiluz you agree with this PR as well?

**Comment by javiereguiluz** [comment]
@Tobion sure! I love it. @peterrehm thanks for adding this feature.

**Comment by Tobion** [comment]
Ok to merge in 2.8 as a usability fix? Or must wait for 3.1?

**Comment by Tobion** [comment]
Tests are failing

**Comment by peterrehm** [comment]
Should be fixed now.

**Comment by Tobion** [comment]
Thank you @peterrehm.

**Pull request merged**